### PR TITLE
Add gcs service discovery to automatically detect GCS server restart

### PR DESCRIPTION
**Pull request by ffbin**
## Why are these changes needed?
Currently, when gcs server is restarted, GCS client needs an RPC request to find that GCS server fails, and then reconnects and resubscribes. Raylet and CoreWorker will use GCS client, raylet will send heartbeat frequently, but CoreWorker will not, so we need the ability of service discovery to ensure that messages subscribed by CoreWorker will not be lost.
In this pr, we add a timer in GCS client to obtain the address of GCS server periodically and judge whether it changes to determine whether GCS server is restarted.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc [hunk 1]
```diff
@@ -117,5 +137,26 @@ void ServiceBasedGcsClient::GetGcsServerAddressFromRedis(
   address->second = std::stoi(result.substr(pos + 1));
 }
 
+void ServiceBasedGcsClient::Tick() {
+  auto address = get_server_address_func_();
+  re_subscribe_func_(false, address);
+  ScheduleTick();
+}
+
+void ServiceBasedGcsClient::ScheduleTick() {
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_gcs_client.cc:145]: Just put `ScheduleTick` in `Tick`?

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc [hunk 2]
```diff
@@ -39,28 +39,32 @@ Status ServiceBasedGcsClient::Connect(boost::asio::io_service &io_service) {
...
-    actor_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
-    node_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
-    task_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
-    object_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
-    worker_accessor_->AsyncResubscribe(is_pubsub_server_restarted);
+  std::pair<std::string, int> address = get_server_address_func_();
+
+  re_subscribe_func_ = [this](bool is_pubsub_server_restarted,
+                              const std::pair<std::string, int> &address) {
+    if (address != current_gcs_server_address_) {
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_gcs_client.cc:55]: Better put this check outside of `re_subscribe_func_`

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc [hunk 3]
```diff
@@ -39,28 +39,32 @@ Status ServiceBasedGcsClient::Connect(boost::asio::io_service &io_service) {
...
   };
 
   // Connect to gcs service.
   client_call_manager_.reset(new rpc::ClientCallManager(io_service));
-  gcs_rpc_client_.reset(new rpc::GcsRpcClient(address.first, address.second,
-                                              *client_call_manager_, get_server_address,
-                                              re_subscribe));
+  gcs_rpc_client_.reset(
+      new rpc::GcsRpcClient(address.first, address.second, *client_call_manager_,
+                            get_server_address_func_, re_subscribe_func_));
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_gcs_client.cc]: I think we no longer need to call `re_subscribe_func_` in `GcsRpcClient`, right?

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc [hunk 4]
```diff
@@ -39,28 +39,32 @@ Status ServiceBasedGcsClient::Connect(boost::asio::io_service &io_service) {
   gcs_pub_sub_.reset(new GcsPubSub(redis_gcs_client_->GetRedisClient()));
 
   // Get gcs service address.
-  auto get_server_address = [this]() {
+  get_server_address_func_ = [this]() {
     std::pair<std::string, int> address;
     GetGcsServerAddressFromRedis(redis_gcs_client_->primary_context()->sync_context(),
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_gcs_client.cc:44]: Now we are calling `GetGcsServerAddressFromRedis` periodically. It shouldn't retry and sleep, otherwise it will block the event loop

File: src/ray/gcs/pubsub/gcs_pub_sub.h [hunk 5]
```diff
@@ -88,6 +88,12 @@ class GcsPubSub {
   /// \return Status
   Status Unsubscribe(const std::string &channel, const std::string &id);
 
+  /// Unsubscribe to messages with the specified channel.
```
Comment (raulchen) [inline src/ray/gcs/pubsub/gcs_pub_sub.h:91]:
```suggestion
  /// Unsubscribe to messages from the specified channel.
```

File: src/ray/gcs/pubsub/gcs_pub_sub.cc [hunk 6]
```diff
@@ -44,17 +44,12 @@ Status GcsPubSub::SubscribeAll(const std::string &channel, const Callback &subsc
...
-  auto channel = channels_.find(pattern);
-  RAY_CHECK(channel != channels_.end());
-  channel->second.command_queue.push_back(Command());
+Status GcsPubSub::Unsubscribe(const std::string &channel, const std::string &id) {
+  return UnsubscribeInternal(channel, id);
+}
 
-  // Process the first command on the queue, if possible.
-  return ExecuteCommandIfPossible(channel->first, channel->second);
+Status GcsPubSub::UnsubscribeAll(const std::string &channel) {
```
Comment (stephanie-wang) [inline src/ray/gcs/pubsub/gcs_pub_sub.cc:51]: Why is this change needed? No one seems to call it.

**Review by stephanie-wang — COMMENTED** [review]
I don't really understand why the client needs to periodically ping the GcsServer. At some point, won't it need to connect a new RPC client to the GcsServer at the changed address anyway? Why can't we just resubscribe then?

Edit: Okay, reread the description and I think I understand now. Can we just have the raylet tell the core workers to resubscribe? I'm a bit worried about having to send all these heartbeat messages when there are no failures.

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc (hunk 2 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc (hunk 1 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc (hunk 3 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc (hunk 4 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.cc]: GetGcsServerAddressFromRedis will be retried only when gcs client is initialized, because we need to wait for GCS server to start.

File: src/ray/gcs/pubsub/gcs_pub_sub.h (hunk 5 above)
Comment (ffbin) [reply src/ray/gcs/pubsub/gcs_pub_sub.h]: fixed

File: src/ray/gcs/pubsub/gcs_pub_sub.cc (hunk 6 above)
Comment (ffbin) [reply src/ray/gcs/pubsub/gcs_pub_sub.cc]: I hava remove this change, thanks.

**Review by stephanie-wang — DISMISSED** [review]
Reposting since I'm not sure if you saw: Can we just have the raylet tell the core workers to resubscribe? I'm a bit worried about having to send all these heartbeat messages when there are no failures.

**Comment by ffbin** [comment]
@stephanie-wang Sorry to reply now, I think this is a good optimization method, which can reduce the access pressure of core worker to redis. I think if I can add a switch to GCS client, raylet's GCS client can open service discovery, but the GCS client of CoreWorker doesn't need to be turned on,    because GCS client is not only used in CoreWorker, such as GlobalStateAccessor.

**Comment by raulchen** [comment]
@stephanie-wang Do you suggest adding a RPC to let raylet inform core workers when it detects GCS server address changes? 
I'm a bit concerned that this would make the logic a bit too complicated.
By the way, the clients do not ping the GCS server. They periodically ping the service discovery server, which is currently the Redis server and might be replaced by zookeeper or etcd in the future. 
I think such workload (e.g., a few thousand QPS for a cluster of 100 nodes) should be fine for Redis.

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 7]
```diff
@@ -241,25 +239,20 @@ class GcsRpcClient {
...
       int index = 0;
       for (; index < RayConfig::instance().ping_gcs_rpc_server_max_retries(); ++index) {
-        address = get_server_address_();
-        RAY_LOG(DEBUG) << "Attempt to reconnect to GCS server: " << address.first << ":"
-                       << address.second;
-        if (Ping(address.first, address.second, 100)) {
-          RAY_LOG(INFO) << "Reconnected to GCS server: " << address.first << ":"
-                        << address.second;
-          break;
+        if (get_server_address_(&address)) {
```
Comment (raulchen) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:242]: Better also move this (recreating the RPC client) to also ServiceBasedGcsClient. So we can handle the RPC stuff and the pub sub stuff in one place.

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc [hunk 8]
```diff
@@ -79,42 +84,75 @@ Status ServiceBasedGcsClient::Connect(boost::asio::io_service &io_service) {
...
-  address->first = result.substr(0, pos);
-  address->second = std::stoi(result.substr(pos + 1));
+  return false;
+}
+
+void ServiceBasedGcsClient::Tick() {
+  std::pair<std::string, int> address;
+  if (get_server_address_func_(&address)) {
+    // In order to avoid repeated subscriptions, we added a check of current gcs
+    // server address.
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.cc:138]: fixed

File: src/ray/gcs/gcs_client/service_based_gcs_client.h [hunk 9]
```diff
@@ -41,8 +41,14 @@ class RAY_EXPORT ServiceBasedGcsClient : public GcsClient {
...
   /// \param address The address of gcs server.
-  void GetGcsServerAddressFromRedis(redisContext *context,
-                                    std::pair<std::string, int> *address);
+  /// \param retry_count The maximum number of retries to get gcs server rpc address.
+  /// \return Returns true if gcs server address is obtained, False otherwise.
+  bool GetGcsServerAddressFromRedis(redisContext *context,
+                                    std::pair<std::string, int> *address,
+                                    int retry_count = 1);
+
+  /// A periodic timer that fires on every gcs server address check period.
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.h]: fixed

File: src/ray/gcs/gcs_client/service_based_gcs_client.h [hunk 10]
```diff
@@ -41,8 +41,14 @@ class RAY_EXPORT ServiceBasedGcsClient : public GcsClient {
   ///
   /// \param context The context of redis.
   /// \param address The address of gcs server.
-  void GetGcsServerAddressFromRedis(redisContext *context,
-                                    std::pair<std::string, int> *address);
+  /// \param retry_count The maximum number of retries to get gcs server rpc address.
+  /// \return Returns true if gcs server address is obtained, False otherwise.
+  bool GetGcsServerAddressFromRedis(redisContext *context,
+                                    std::pair<std::string, int> *address,
+                                    int retry_count = 1);
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.h]: fixed

File: src/ray/rpc/gcs_server/gcs_rpc_client.h (hunk 7 above)
Comment (ffbin) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h]: fixed

**Comment by rkooo567** [comment]
This PR is missing any test. What's the test plan? Are you guys trying to make a huge test PR in the future to verify all restarting logic works properly?

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.cc:138]: fixed

File: src/ray/gcs/gcs_client/service_based_gcs_client.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.h]: fixed

File: src/ray/gcs/gcs_client/service_based_gcs_client.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.h]: fixed

File: src/ray/rpc/gcs_server/gcs_rpc_client.h (hunk 7 above)
Comment (ffbin) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h]: fixed

**Comment by ffbin** [comment]
@rkooo567 Hi, service_based_gcs_client_test.cc has test cases for each table, and they will test the service discovery scenario. Previously, it was necessary to send an RPC request to trigger the re subscription. Now it is not needed, so I modified the TestActorTableResubscribe test case.

**Review by raulchen — APPROVED** [review]

**Comment by raulchen** [comment]
@mehrdadn Windows CI failed. But it doesn't look related. So I merged this. Please let us know if you think it's related.

**Comment by mehrdadn** [comment]
@raulchen Yeah it's not related, thanks for checking. Would you mind merging #9173 actually? It should get rid of that flaky test.
